dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Inlining.java

```diff
@@ -0,0 +1,187 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.exception.MalformedProgramException;
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.processing.ExprTransformer;
+import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.Thread;
+import com.dat3m.dartagnan.program.event.EventFactory;
+import com.dat3m.dartagnan.program.event.EventUser;
+import com.dat3m.dartagnan.program.event.Tag;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.event.core.Label;
+import com.dat3m.dartagnan.program.event.core.utils.RegReader;
+import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
+import com.dat3m.dartagnan.program.event.functions.AbortIf;
+import com.dat3m.dartagnan.program.event.functions.DirectFunctionCall;
+import com.dat3m.dartagnan.program.event.functions.DirectValueFunctionCall;
+import com.dat3m.dartagnan.program.event.functions.Return;
+import com.dat3m.dartagnan.program.event.lang.llvm.LlvmCmpXchg;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.configuration.Option;
+import org.sosy_lab.common.configuration.Options;
+
+import java.util.*;
+
+import static com.dat3m.dartagnan.configuration.OptionNames.RECURSION_BOUND;
+import static com.dat3m.dartagnan.program.event.EventFactory.*;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@Options
+public class Inlining implements ProgramProcessor {
+
+    @Option(name = RECURSION_BOUND,
+            description = "Inlines each function call up to this many times.",
+            secure = true)
+    private int bound = 1;
+
+    private Inlining() {}
+
+    public static Inlining fromConfig(Configuration config) throws InvalidConfigurationException {
+        Inlining process = new Inlining();
+        config.inject(process);
+        return process;
+    }
+
+    @Override
+    public void run(Program program) {
+        for (Function function : program.getFunctions()) {
+            inlineAllCalls(function);
+        }
+    }
+
+    private void inlineAllCalls(Function thread) {
```

Comment:
is the parameter a thread or a function?

Reply:
A function now. Will rename